File: workspace/2023-02-25/try-prolog.prolog/fizz_buzz.pl

query_example :-
    fizz_buzz(100).

fizz_buzz(N) :-
    fizz_buzz_range(1, N).

fizz_buzz_range(Min, Max) :-
    Min =< Max,
    fizz_buzz_value(Min),
    is(NextMin, Min + 1),
    fizz_buzz_range(NextMin, Max).

fizz_buzz_value(N) :-
    0 is N mod 15, write('FizzBuzz'), nl, !.
fizz_buzz_value(N) :-
    0 is N mod 5, write('Buzz'), nl, !.
fizz_buzz_value(N) :-
    0 is N mod 3, write('Fizz'), nl, !.
fizz_buzz_value(N) :-
    write(N), nl, !.

File: workspace/2023/2023-01-25/fizz-buzz.forth/fizz-buzz.fs

: fizz_buzz_value ( n -- )
    dup 15 mod 0 = if
        s" FizzBuzz " type drop
    else
        dup 5 mod 0 = if
            s" Buzz " type drop
        else
            dup 3 mod 0 = if
                s" Fizz " type drop
            else
                .
            endif
        endif
    endif
;

: _fizz_buzz recursive ( max min -- )
    over over < if
        drop drop
    else
        dup fizz_buzz_value
        1 +
        _fizz_buzz
    endif
;

: fizz_buzz ( n -- )
    1
    _fizz_buzz
;

100 fizz_buzz

